### PR TITLE
ofs_cpeng: first version of ofs_cpeng umd

### PR DESCRIPTION
**Pull request by r-rojo**
* Fix ofs_parse.py
  * Only assume the first argument to an ast.Call object is of ast.Name if the func.id is ofs_addr.
* add wait macros to libofs header
  * Use same algorighm in ofs_wait_for_eq function to define macros to wait for bitfields.
    This is done because passing a reference to a bitfield to a function results in a compile error/warning.
* Add .yml file that describes the register map of the cpeng feature.
  * Include one driver function "wait_for_ready"
* Add unit tests for wait_for_ready

### CODE REVIEW

File: ofs/libofs/ofs_primitives.h [hunk 1]
```diff
@@ -44,6 +44,47 @@
...
+        _status;                                                            \
+})                                                                          \
+
+#define OFS_WAIT_FOR_NE(_bit, _value, _timeout_usec, _sleep_usec)           \
+({                                                                          \
+	int _status = 0;                                                    \
+	OFS_TIMESPEC_USEC(ts, _sleep_usec);                                 \
+	struct timespec begin, now;                                         \
+	clock_gettime(CLOCK_MONOTONIC, &begin);                             \
+	while(_bit != _value) {                                             \
```
Comment (hundeboll-silicom) [inline ofs/libofs/ofs_primitives.h:74]: I would expect this comparison to be `_bit == _value` ?

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -40,3 +40,4 @@ add_subdirectory(userclk)
 add_subdirectory(fpgametrics)
 add_subdirectory(libofs)
 add_subdirectory(ofs_driver)
+add_subdirectory(ofs_cpeng)
```
Comment (hundeboll-silicom) [inline tests/CMakeLists.txt:43]: Does this test compile/pass if `OPAE_BUILD_OFS_CPENG` is not true?

File: ofs/drivers/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+## IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+## ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+## LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+## CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+## SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+## INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+## CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+## ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+## POSSIBILITY OF SUCH DAMAGE.
+option(OPAE_BUILD_OFS_CPENG "Enable building ofs_cpeng driver" ON)
```
Comment (hundeboll-silicom) [inline ofs/drivers/CMakeLists.txt:26]: Empty line between license text and code (goes for the all files).

File: ofs/drivers/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+mark_as_advanced(OPAE_BUILD_OFS_CPENG)
+
+if (OPAE_BUILD_OFS_CPENG)
+    add_subdirectory(cpeng)
+endif (OPAE_BUILD_OFS_CPENG)
+
```
Comment (hundeboll-silicom) [inline ofs/drivers/CMakeLists.txt:32]: No empty line(s) at end of file (same for all files).

File: ofs/libofs/ofs_primitives.h (hunk 1 above)
Comment (r-rojo) [reply ofs/libofs/ofs_primitives.h]: You're right, I had it right the first time. It should be fixed now. Thanks 😄

File: tests/CMakeLists.txt (hunk 2 above)
Comment (r-rojo) [reply tests/CMakeLists.txt:43]: Probably not, I'll fix that.

File: tests/ofs_cpeng/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+## IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+## ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+## LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+## CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+## SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+## INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+## CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+## ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+## POSSIBILITY OF SUCH DAMAGE.
+opae_test_add(
```
Comment (tswhison) [inline tests/ofs_cpeng/CMakeLists.txt:26]: Need a blank here, too.

**Review by tswhison — APPROVED** [review]

**Pull request merged**